src/aquarium-optimized/Aquarium.cpp

```diff
@@ -780,46 +780,53 @@ void Aquarium::render() {
       resetFpsTime();
     }
 
-  bool updateAndDrawForEachFish =
-      toggleBitset.test(static_cast<size_t>(TOGGLE::UPATEANDDRAWFOREACHMODEL));
-
-  if (updateAndDrawForEachFish) {
-    updateAndDrawBackground();
-    updateAndDrawFishes();
-    mContext->updateFPS(mFpsTimer, &mCurFishCount, &toggleBitset);
-  } else {
-    updateBackground();
-    updateFishes();
-    mContext->updateFPS(mFpsTimer, &mCurFishCount, &toggleBitset);
+  mDrawPerModel = toggleBitset.test(static_cast<size_t>(TOGGLE::DRAWPERMODEL));
 
-    // Begin render pass
+  updateBackground();
+  updateFishes();
+  mContext->updateFPS(mFpsTimer, &mCurFishCount, &toggleBitset);
+  if (!mDrawPerModel) {
     mContext->beginRenderPass();
-
     drawBackground();
     drawFishes();
     mContext->showFPS();
-
-    // End renderpass
   }
 }
 
-void Aquarium::updateAndDrawBackground() {
+void Aquarium::updateBackground() {
   Model *model = mAquariumModels[MODELNAME::MODELRUINCOlOMN];
   for (int i = MODELNAME::MODELRUINCOlOMN; i <= MODELNAME::MODELSEAWEEDB; ++i) {
     model = mAquariumModels[i];
-    updateWorldMatrixAndDraw(model);
+    updateWorldMatrix(model);
   }
 }
 
-void Aquarium::drawBackground() {
-  Model *model = mAquariumModels[MODELNAME::MODELRUINCOlOMN];
-  for (int i = MODELNAME::MODELRUINCOlOMN; i <= MODELNAME::MODELSEAWEEDB; ++i) {
-    model = mAquariumModels[i];
-    model->draw();
+void Aquarium::updateWorldMatrix(Model *model) {
+  if (model->worldmatrices.size()) {
+    for (auto &world : model->worldmatrices) {
+      updateWorldProjections(world);
+      if (mDrawPerModel)
+        model->prepareForDraw();
+      model->updatePerInstanceUniforms(worldUniforms);
+      if (mDrawPerModel)
+        model->draw();
+    }
   }
+  if (!mDrawPerModel)
+    model->prepareForDraw();
 }
 
-void Aquarium::updateAndDrawFishes() {
+void Aquarium::updateWorldProjections(const std::vector<float> &w) {
+  ASSERT(w.size() == 16);
+  memcpy(worldUniforms.world, w.data(), 16 * sizeof(float));
+  matrix::mulMatrixMatrix4(worldUniforms.worldViewProjection,
+                           worldUniforms.world,
+                           lightWorldPositionUniform.viewProjection);
+  matrix::inverse4(g.worldInverse, worldUniforms.world);
+  matrix::transpose4(worldUniforms.worldInverseTranspose, g.worldInverse);
+}
+
+void Aquarium::updateFishes() {
```

Comment:
The function name seems a little confusing now, as it means "update & draw" in some cases.